src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -672,71 +678,73 @@ const static auto formats = {
     std::locale(std::locale::classic(),
                 new boost::local_time::local_time_input_facet("%Y-%m-%d %H:%M:%s%ZP")),
 };
-const static boost::posix_time::ptime epoch(boost::gregorian::date(1970, 1, 1));
-const static boost::posix_time::ptime max_date(boost::gregorian::date(2030, 1, 1));
 
-static std::chrono::milliseconds parseDateAsMillis(
-    const std::string& datetime, const boost::posix_time::ptime& default_time = epoch) {
-    boost::posix_time::ptime utc_time{boost::posix_time::not_a_date_time};
-    auto millis = 0LL;
+class DateToIntGenerator : public Generator<int64_t> {
+public:
+    explicit DateToIntGenerator(UniqueGenerator<bsoncxx::types::b_date>& generator)
+        : _generator{std::move(generator)} {}
+    int64_t evaluate() override {
+        auto date = _generator->evaluate();
+        return date.to_int64();
+    }
 
-    // Empty datetime is 0.
-    if (!datetime.empty()) {
-        for (const auto& format : formats) {
-            std::istringstream date_stream{datetime};
-            date_stream.imbue(format);
-            boost::local_time::local_date_time local_date{boost::local_time::not_a_date_time};
-            if (date_stream >> local_date) {
-                utc_time = local_date.utc_time();
+private:
+    const UniqueGenerator<bsoncxx::types::b_date> _generator;
+};
 
-                millis = (utc_time - epoch).total_milliseconds();
-                break;
-            }
-        }
+class CastDoubleToInt64Generator : public Generator<int64_t> {
+public:
+    explicit CastDoubleToInt64Generator(UniqueGenerator<double>& generator)
+        : _generator{std::move(generator)} {}
+    int64_t evaluate() override {
+        return (int64_t)_generator->evaluate();
+    }
 
-        if (utc_time == boost::posix_time::not_a_date_time) {
-            try {
-                // Last gasp try to interpret unsigned long long.
-                millis = std::stoull(datetime);
-            } catch (const std::invalid_argument& _) {
-                auto msg = "^RandomDate: Invalid Dateformat '" + datetime + "'";
-                BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax{msg});
-            }
-        }
+private:
+    const UniqueGenerator<double> _generator;
+};
 
-    } else {
-        millis = (default_time - epoch).total_milliseconds();
+class DateStringParserGenerator : public Generator<int64_t> {
+public:
+    explicit DateStringParserGenerator(UniqueGenerator<std::string>& generator)
+        : _generator{std::move(generator)} {}
```

Comment:
I think this can introduce use-after-move problems:

```c++
UniqueGenerator<std::string> og = ....;
DateStringParserGenerator other {og};
og->evaluate(); // 💥
```

I realize mostly academic for this code, but see if you can just pass in an rvalue and do the `std::move` at the callsite:

```c++
class ... {
     explicit DateStringParserGenerator(UniqueGenerator<std::string>&& generator)
     : _gen{generator} {}
}
...

UniqueGenerator<std::string> og = ....;
DateStringParserGenerator other {std::move(of)};
og->evaluate(); // still runtime 💥 but compiler can detect it
```

Ugh my goodness C++ makes it far too easy to footgun.

Reply:
Thanks, I'll change it. I did think of doing the move at the call site, but thought it would be easier to do it in the receiver.

Reply:
It looks like there is some oddness in my linux dev environment, the code above fails to compile on ubuntu (it works on MacOS (thanks @rtimmons). I will investigate my dev env a bit more and come back.